src/java/org/apache/cassandra/index/sai/disk/v2/V2VectorIndexSearcher.java

```diff
@@ -195,21 +203,36 @@ private BitsOrPostingList bitsOrPostingListForKeyRange(QueryContext context, Abs
             // If num of matches are not bigger than limit, skip ANN.
             // (nRows should not include shadowed rows, but context doesn't break those out by segment,
             // so we will live with the inaccuracy.)
-            var nRows = Math.toIntExact(maxSSTableRowId - minSSTableRowId + 1);
+            int nRows = Math.toIntExact(maxSSTableRowId - minSSTableRowId + 1);
             int maxBruteForceRows = min(globalBruteForceRows, maxBruteForceRows(limit, nRows, graph.size()));
-            logger.trace("Search range covers {} rows; max brute force rows is {} for sstable index with {} nodes, LIMIT {}",
+            if (logger.isTraceEnabled())
+                logger.trace("Search range covers {} rows; max brute force rows is {} for sstable index with {} nodes, LIMIT {}",
                          nRows, maxBruteForceRows, graph.size(), limit);
-            Tracing.trace("Search range covers {} rows; max brute force rows is {} for sstable index with {} nodes, LIMIT {}",
+            if (Tracing.isTracing())
+                Tracing.trace("Search range covers {} rows; max brute force rows is {} for sstable index with {} nodes, LIMIT {}",
                           nRows, maxBruteForceRows, graph.size(), limit);
+
             if (nRows <= maxBruteForceRows)
             {
-                IntArrayList postings = new IntArrayList(Math.toIntExact(nRows), -1);
-                for (long sstableRowId = minSSTableRowId; sstableRowId <= maxSSTableRowId; sstableRowId++)
+                var segmentRowIdsStream = LongStream.range(minSSTableRowId, maxSSTableRowId + 1)
+                                                    .filter(sstableRowId -> context.shouldInclude(sstableRowId, primaryKeyMap))
+                                                    .mapToInt(metadata::toSegmentRowId);
+                final int[] postings;
+                if (graph.getCompressedVectors() == null || nRows <= limit)
                 {
-                    if (context.shouldInclude(sstableRowId, primaryKeyMap))
-                        postings.addInt(metadata.toSegmentRowId(sstableRowId));
+                    IntArrayList arrayList = new IntArrayList(nRows, -1);
```

Comment:
can't we just call IntStream.toArray directly?

Reply:
of course.